### PR TITLE
feat(indexer): Enable devnet indexer as a preset

### PR DESCRIPTION
**Pull request by roninjin10**
- enable devnet as a preset
- read from devnet artifacts to get devnet addresses

### CODE REVIEW

**Comment by roninjin10** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #7271** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7271" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/7271?utm_source=stack-comment).

File: indexer/config/devnet.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type ethereumContracts struct {
```
Comment (hamdiallam) [inline indexer/config/devnet.go]: does parsing fail if you leave out the unneeded fields? If we simply supply `json` tags to the `config.L1Contracts` structs can we deserialize directly into that?

**Review by hamdiallam — COMMENTED** [review]
Sweet, this is clean. Left a comment about deserializing directly into `L1Contracts`

File: indexer/config/devnet.go (hunk 1 above)
Comment (roninjin10) [reply indexer/config/devnet.go]: Hmm curious too. It's a common.Address not a string so my guess is no but curius if it works just from setting type to common.Address

Comment (roninjin10) [reply indexer/config/devnet.go]: My guess is yes we can leave out the unneeded fields but will give that a try to

File: indexer/config/devnet.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func readEthereumContracts(filename string) (*ethereumContracts, error) {
+	// Check if the file exists
+	if _, err := os.Stat(filename); os.IsNotExist(err) {
+		return nil, err
+	}
+
+	// Read the file
+	content, err := ioutil.ReadFile(filename)
```
[the remark this answers is not in the document]
Comment (roninjin10) [reply indexer/config/devnet.go]: AI's fault not my fault will fix

File: indexer/config/devnet.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the JSON
```
Comment (roninjin10) [inline indexer/config/devnet.go]:
```suggestion
```
remove copilot comment

File: indexer/config/devnet.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	var contracts ethereumContracts
+	if err := json.Unmarshal(content, &contracts); err != nil {
+		return nil, err
+	}
+
+	return &contracts, nil
+}
+
+func GetDevnetPreset() (*Preset, error) {
+	ethereumContracts, err := readEthereumContracts("../.devnet/addresses.json")
```
Comment (roninjin10) [inline indexer/config/devnet.go]: move this path to a constant at top of the file

File: indexer/config/devnet.go (hunk 1 above)
Comment (hamdiallam) [reply indexer/config/devnet.go]: Just checked and `common.Address` implements the `UnmarshalJSON` interface so it should work!

Comment (roninjin10) [reply indexer/config/devnet.go]: nice TIL about UnmarshalJSON

File: indexer/config/config.go [hunk 5]
```diff
@@ -157,18 +157,21 @@ func LoadConfig(log log.Logger, path string) (Config, error) {
...
 		}
-
 		log.Info("detected preset", "preset", conf.Chain.Preset, "name", preset.Name)
 		log.Info("setting L1 information from preset")
-		conf.Chain.L1Contracts = preset.ChainConfig.L1Contracts
-		conf.Chain.L1StartingHeight = preset.ChainConfig.L1StartingHeight
-		conf.Chain.L1BedrockStartingHeight = preset.ChainConfig.L1BedrockStartingHeight
-		conf.Chain.L2BedrockStartingHeight = preset.ChainConfig.L1BedrockStartingHeight
+		conf.Chain = preset.ChainConfig
+		log.Info(conf.Chain.L1Contracts.AddressManager.String())
```
Comment (hamdiallam) [inline indexer/config/config.go]:
```suggestion
```

File: indexer/config/devnet.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return &Preset{
+		Name: "devnet",
+		ChainConfig: ChainConfig{
+			Preset:                  DEVNET_L2_CHAIN_ID,
+			L1StartingHeight:        0,
+			L1BedrockStartingHeight: 0,
+			L2BedrockStartingHeight: 0,
```
Comment (hamdiallam) [inline indexer/config/devnet.go]:
```suggestion
```

Since the default value for these fields are 0, you don't have to explicitly set them

**Review by hamdiallam — APPROVED** [review]
LGTM just two small nits

**Pull request merged**